rust/src/mapmanip/procgen/mod.rs

```diff
@@ -0,0 +1,565 @@
+///     "hallway_vertical_marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/hallway_vertical",
+///     "hallway_node_marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/hallway_floor",
+///     "room_configs": [
+///         {
+///             "marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/small",
+///             "width": 3,
+///             "height": 3,
+///             "min": 7,
+///             "max": 9
+///         },
+///         {
+///             "marker": "/obj/effect/map_effect/marker/mapmanip/submap/insert/deepmaints/medium",
+///             "width": 5,
+///             "height": 5,
+///             "min": 2,
+///             "max": 3
+///         }
+///     ]
+/// }
+/// ```
+///
+/// ## Shortcomings
+///
+/// 1. Currently this mapmanip only works on the 1st z-level of a map file. This
+///    isn't an impossible thing to fix, I just didn't feel like setting it up
+///    to handle other z-levels.
+/// 2. As a result of the mazegen algorithm using odd-sized rooms to ensure
+///    consistent spacing for hallways and junctions, and the fact that the maze
+///    is generated on a grid scaled to 1/3rd the original map size, all rooms
+///    must have sides with dimensions that are both odd and multiples of three,
+///    e.g. 9, 15, 21.
+struct MazegenHauberk {
+    settings: MazegenHauberkSettings,
+    grid: CellGrid,
+    width: i32,
+    height: i32,
+    regions: CellGrid,
+    rooms: Vec<Rect>,
+    current_region: i32,
+}
+
+impl MazegenHauberk {
+    fn new(settings: MazegenHauberkSettings, width: i32, height: i32) -> Self {
+        MazegenHauberk {
+            settings,
+            width,
+            height,
+
+            grid: vec![vec![MAP_TILE_WALL; height as usize]; width as usize],
+            regions: vec![vec![-1; height as usize]; width as usize],
+            rooms: vec![],
+            current_region: -1,
+        }
+    }
+
+    fn can_place_room(&self, room: &Rect) -> bool {
+        if self.rooms.iter().any(|other| room.intersects_with(other)) {
+            return false;
+        }
+
+        for a in room.x1..room.x2 {
+            for b in room.y1..room.y2 {
+                let val = self.grid[a as usize][b as usize];
+                if val == MAP_TILE_IGNORE || val >= 0 {
+                    return false;
+                }
+            }
+        }
+
+        true
+    }
+
+    fn add_rooms(&mut self, rng: &mut impl Rng) {
+        let mut room_config_idxes = vec![];
+        for (idx, room_config) in self.settings.room_configs.iter().enumerate() {
+            let room_type_count = rng.gen_range(room_config.min..=room_config.max);
+            for _ in 0..room_type_count {
+                room_config_idxes.push(idx);
+            }
+        }
+
+        room_config_idxes.shuffle(rng);
+        for _ in 0..=200 {
```

Comment:
is it possible that the length of room_config_idxes might be greater than 200? Should this check account for this?